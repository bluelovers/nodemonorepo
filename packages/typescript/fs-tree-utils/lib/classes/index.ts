import FileSystemRepresentation from './fs-representation'
export {FileSystemRepresentation}
