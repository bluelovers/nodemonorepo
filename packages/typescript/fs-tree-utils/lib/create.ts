import {WriteTree} from './types'
import FileSystemRepresentation from './classes/fs-representation'
const {File, Directory} = FileSystemRepresentation

/**
 * Create a directory tree
 * @param tree Tree structure that needs to create
 * @param container Where to place the tree
 */
export async function create (tree: WriteTree, container: string = ''): Promise<void> {
  if (typeof tree === 'string') {
    return create(new File(tree), container)
  }

  if (typeof tree === 'function') {
    return tree(container)
  }

  if (tree instanceof FileSystemRepresentation) {
    return create(x => tree.write(x), container)
  }

  return create(new Directory(tree), container)
}

export namespace create {
  export const async = create
}

export default create
