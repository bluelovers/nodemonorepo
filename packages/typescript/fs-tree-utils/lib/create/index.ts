import * as types from '../.types'

/**
 * Create a directory tree
 * @param tree Tree structure that needs to create
 * @param container Where to place the tree
 */
export async function create (
  tree: types.Tree.Write,
  container: string
): Promise<void> {
  const { FileSystemRepresentation } = types

  if (typeof tree === 'string' || tree instanceof Buffer) {
    return create(new FileSystemRepresentation.File(tree), container)
  }

  if (typeof tree === 'function') {
    return tree(container, { create })
  }

  if (tree instanceof FileSystemRepresentation) {
    return create(x => tree.write(x, { create }), container)
  }

  return create(new FileSystemRepresentation.Directory(tree), container)
}

export namespace create {
  export const async = create
}

export default create
