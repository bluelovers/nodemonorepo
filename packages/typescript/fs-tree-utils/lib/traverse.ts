import * as path from 'path'
import * as fsx from 'fs-extra'

export type TraversalDeepFuncParam = {
  container: string,
  item: string,
  path: string,
  stats: fsx.Stats,
  level: number
}
export type DeepFunc = (x: TraversalDeepFuncParam) => boolean
export type DepsDict = {
  [packageName: string]: string
}
export type TraversalResultItem = TraversalDeepFuncParam
export type TraversalResult = TraversalResultItem[]
export type AsyncTraversalResult = Promise<TraversalResult>

export async function traverse (
  dirname: string,
  deep: DeepFunc = () => true,
  level = 0
): AsyncTraversalResult {
  const dirChildren = await fsx.readdir(dirname)
  let result: TraversalResult = []

  for (const item of dirChildren) {
    const itemPath = path.join(dirname, item)
    const stats = fsx.statSync(itemPath)
    const itemResult = {
      item,
      stats,
      level,
      container: dirname,
      path: itemPath
    }
    result.push(itemResult)

    const shouldGoDeeper = stats.isDirectory() && deep(itemResult)
    if (shouldGoDeeper) {
      result.push(...await traverse(itemPath, deep, level + 1))
    }
  }

  return result
}

export default Object.assign(
  traverse,
  {
    async: traverse
  }
)
