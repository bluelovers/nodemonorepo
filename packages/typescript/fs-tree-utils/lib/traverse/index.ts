import * as path from 'path'
import * as fsx from 'fs-extra'
import {Traverse} from '../.types'

/**
 * @param dirname Top directory
 * @param options
 * @param options.deep When to dive deeper
 * @param options.stat Stat function to use
 * @param options.level Initial level of depth
 * @returns Array of traversed files/directories
 */
export function traverse (
  dirname: string,
  options: Traverse.Options = {}
): Traverse.Result {
  const {
    deep = () => true,
    stat = (x: string) => fsx.stat(x),
    level = 0
  } = options

  return main(dirname, level)

  async function main (
    dirname: string,
    level: Traverse.Options.Level
  ): Traverse.Result {
    const dirChildren = await fsx.readdir(dirname)
    const result = Array<Traverse.Result.Item>()

    const statList = await Promise.all(
      dirChildren.map(async item => {
        const itemPath = path.join(dirname, item)
        const stats = await Promise.resolve(stat(itemPath))
        return {item, itemPath, stats}
      })
    )

    for (const {item, itemPath, stats} of statList) {
      const itemResult = {
        item,
        stats,
        level,
        container: dirname,
        path: itemPath
      }
      result.push(itemResult)

      const shouldGoDeeper = stats.isDirectory() && deep(itemResult)
      if (shouldGoDeeper) {
        result.push(...await main(itemPath, level + 1))
      }
    }

    return result
  }
}

export default traverse
