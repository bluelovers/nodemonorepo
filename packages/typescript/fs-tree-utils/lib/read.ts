import * as path from 'path'
import ramda from 'ramda'
import * as fsx from 'fs-extra'
import traverse, {DeepFunc} from './traverse'
import {Tree, TreeObject, FileContent} from './types'

export type NestedReadResult = Promise<Tree>

export type FlatReadResultFileContent = {[filename: string]: FileContent}

export type FlatReadResultValue = {
  fileContents: FlatReadResultFileContent,
  directories: string[],
  files: string[],
  all: string[]
}

export type FlatReadResult = Promise<FlatReadResultValue>

/**
 * @param name Directory name
 * @returns Nested directory tree representation
 */
export async function readNested (name: string): NestedReadResult {
  const stats = await fsx.stat(name)

  if (stats.isFile()) {
    return await fsx.readFile(name, 'utf8')
  }

  if (stats.isDirectory()) {
    let tree: TreeObject = {}
    for (const item of await fsx.readdir(name)) {
      const subtree = await readNested(path.join(name, item))
      tree[item] = subtree
    }
    return tree
  }

  throw new Error(`Unknown filesystem type of path '${name}'`)
}

/**
 * @param name Directory name
 * @param deep When to dive deeper
 * @returns List of files, directories and file contents
 */
export async function readFlat (name: string, deep?: DeepFunc): FlatReadResult {
  const array = await traverse(name, deep)
  const [fileList, dirList] = ramda.partition(x => x.stats.isFile(), array)

  const map = await Promise.all(
    fileList.map(
      (x): [string, Promise<string>] =>
        [x.path, fsx.readFile(x.path, 'utf8')]
    )
  )

  let fileContents: FlatReadResultFileContent = {}
  for (const [name, promise] of map) {
    fileContents[name] = await promise
  }

  return (() => {
    const directories = dirList.map(x => x.path)
    const files = fileList.map(x => x.path)
    const all = [...directories, ...files].sort((a, b) => a < b ? -1 : 1)

    return {
      fileContents,
      directories,
      files,
      all
    }
  })()
}

function addAsyncProperty<X> (x: X) {
  return Object.assign(x, {async: x})
}

export default {
  nested: addAsyncProperty(readNested),
  flat: addAsyncProperty(readFlat)
}
