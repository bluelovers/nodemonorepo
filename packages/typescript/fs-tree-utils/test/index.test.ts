import * as path from 'path'
import {cwd, chdir} from 'process'
import * as fsx from 'fs-extra'
import tempPath from 'unique-temp-path'
import * as subject from '../index'
import {DeepFunc} from '../lib/traverse'
const tree = require('./data/tree.yaml')
const oldCwd = cwd()
const tmpContainer = tempPath('fs-tree-utils.')
const tmp = 'tmp'
const {FileSystemRepresentation} = subject
const {File, Directory} = FileSystemRepresentation

const createTreeGetter = (container: string) => () => Promise.all([
  subject.read.nested(container),
  subject.read.flat(container)
]).then(([nested, flat]) => ({nested, flat}))

beforeEach(async () => {
  await fsx.mkdir(tmpContainer)
  chdir(tmpContainer)
  await fsx.remove(tmp)
  await fsx.mkdir(tmp)
})

afterEach(async () => {
  chdir(oldCwd)
  await fsx.remove(tmpContainer)
})

describe('create function', () => {
  it('creates files/directories that yet to exist', async () => {
    const container = path.join(tmp, 'create.0')
    const getTree = createTreeGetter(container)
    await fsx.remove(container)

    await subject.create(tree, container)
    expect(await getTree()).toMatchSnapshot()
  })

  it('ignores already existed files/directories that is not mismatched', async () => {
    const container = path.join(tmp, 'create.1')
    const getTree = createTreeGetter(container)
    await fsx.remove(container)

    const firstBornFiles = {
      foo: 'foo',
      bar: 'bar',
      baz: new File('baz'),
      qux: new File(Buffer.from('qux'))
    }

    const secondBornFiles = {
      newFileA: 'First New File',
      newFileB: 'Second New File',
      newFileC: Buffer.from('Third New File'),
      newFileD: (x: string) => fsx.writeFile(x, 'Forth New File'),
      newFileE: (x: string) => fsx.writeFileSync(x, 'Fifth New File')
    }

    const firstBornFolders = {
      emptyFoo: {},
      emptyBar: {},
      emptyBaz: new Directory(),
      emptyQux: new Directory({}),
      firstBornFiles
    }

    const secondBornFolders = {
      newTreeA: {},
      newTreeB: {
        empty: {}
      },
      newTreeC: {
        empty: {},
        emptyFile: {
          file: ''
        },
        nonEmptyFile: {
          file: 'Not empty'
        }
      },
      newTreeD: new Directory({
        newTreeA: {},
        newTreeB: new Directory({
          empty: {}
        }),
        newTreeC: {
          empty: {},
          emptyFile: new Directory({
            file: ''
          }),
          nonEmptyFile: {
            file: 'Not empty'
          }
        }
      })
    }

    await subject.create({
      ...firstBornFiles,
      ...firstBornFolders
    }, container)
    const firstBorn = await getTree()

    await subject.create({
      ...firstBornFiles,
      ...firstBornFolders,
      ...secondBornFiles,
      ...secondBornFolders
    }, container)
    const secondBorn = await getTree()

    expect({firstBorn, secondBorn}).toMatchSnapshot()
  })

  describe('throws error when encounter mismatched entities', () => {
    it('in which a file is requested in place of a folder', async () => {
      const container = path.join(tmp, 'create.2.0')
      await fsx.remove(container)

      const existingTree = {
        mismatched: {}
      }

      const expectedTree = {
        mismatched: ''
      }

      await subject.create(existingTree, container)
      await expect(subject.create(expectedTree, container)).rejects.toMatchSnapshot()
    })

    it('in which a folder is requested in place of a file', async () => {
      const container = path.join(tmp, 'create.2.1')
      await fsx.remove(container)

      const existingTree = {
        mismatched: ''
      }

      const expectedTree = {
        mismatched: {}
      }

      await subject.create(existingTree, container)
      await expect(subject.create(expectedTree, container)).rejects.toMatchSnapshot()
    })
  })
})

describe('traverse function', () => {
  const container = path.join(tmp, 'traverse.0')

  const init = async () => {
    await fsx.remove(container)
    await subject.create(tree, container)
  }

  const createFunc = (deep?: DeepFunc, level?: number) => async () =>
    (await subject.traverse(container, deep, level))
      .map(({item, path, container, level}) => ({item, path, container, level}))

  const createTester = (deep?: DeepFunc, level?: number) => async () => {
    const fn = createFunc(deep, level)
    await init()
    expect(await fn()).toMatchSnapshot()
  }

  it('works with default `deep` and `level`', createTester())
  it('only dig names end with "B"', createTester(x => /B$/.test(x.item)))
  it('does not dig names end with "B"', createTester(x => !/B$/.test(x.item)))
  it('works with provided `deep` and `level`', createTester(() => true, 3))
})
