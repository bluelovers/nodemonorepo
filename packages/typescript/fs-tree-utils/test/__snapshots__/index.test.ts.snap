// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create function creates files/directories that yet to exist 1`] = `
Object {
  "flat": Object {
    "all": Array [
      "tmp/create.0/topA",
      "tmp/create.0/topB",
      "tmp/create.0/topB/middleA",
      "tmp/create.0/topB/middleB",
      "tmp/create.0/topB/middleB/bottomA",
      "tmp/create.0/topB/middleB/bottomB",
      "tmp/create.0/topB/middleC",
      "tmp/create.0/topB/middleC/bottomA",
      "tmp/create.0/topB/middleC/bottomB",
      "tmp/create.0/topC",
      "tmp/create.0/topC/middleA",
      "tmp/create.0/topC/middleA/bottomA",
      "tmp/create.0/topC/middleA/bottomB",
      "tmp/create.0/topC/middleB",
      "tmp/create.0/topC/middleB/bottomA",
      "tmp/create.0/topC/middleB/bottomB",
      "tmp/create.0/topC/middleC",
      "tmp/create.0/topC/middleC/bottomA",
      "tmp/create.0/topC/middleC/bottomB",
    ],
    "directories": Array [
      "tmp/create.0/topA",
      "tmp/create.0/topB",
      "tmp/create.0/topB/middleA",
      "tmp/create.0/topB/middleB",
      "tmp/create.0/topB/middleB/bottomA",
      "tmp/create.0/topB/middleC",
      "tmp/create.0/topC",
      "tmp/create.0/topC/middleA",
      "tmp/create.0/topC/middleB",
      "tmp/create.0/topC/middleC",
    ],
    "fileContents": Object {
      "tmp/create.0/topB/middleB/bottomB": "B.B.B",
      "tmp/create.0/topB/middleC/bottomA": "B.C.A",
      "tmp/create.0/topB/middleC/bottomB": "B.C.B",
      "tmp/create.0/topC/middleA/bottomA": "C.A.A",
      "tmp/create.0/topC/middleA/bottomB": "C.A.B",
      "tmp/create.0/topC/middleB/bottomA": "C.B.A",
      "tmp/create.0/topC/middleB/bottomB": "C.B.B",
      "tmp/create.0/topC/middleC/bottomA": "C.C.A",
      "tmp/create.0/topC/middleC/bottomB": "C.C.B",
    },
    "files": Array [
      "tmp/create.0/topB/middleB/bottomB",
      "tmp/create.0/topB/middleC/bottomA",
      "tmp/create.0/topB/middleC/bottomB",
      "tmp/create.0/topC/middleA/bottomA",
      "tmp/create.0/topC/middleA/bottomB",
      "tmp/create.0/topC/middleB/bottomA",
      "tmp/create.0/topC/middleB/bottomB",
      "tmp/create.0/topC/middleC/bottomA",
      "tmp/create.0/topC/middleC/bottomB",
    ],
  },
  "nested": Object {
    "topA": Object {},
    "topB": Object {
      "middleA": Object {},
      "middleB": Object {
        "bottomA": Object {},
        "bottomB": "B.B.B",
      },
      "middleC": Object {
        "bottomA": "B.C.A",
        "bottomB": "B.C.B",
      },
    },
    "topC": Object {
      "middleA": Object {
        "bottomA": "C.A.A",
        "bottomB": "C.A.B",
      },
      "middleB": Object {
        "bottomA": "C.B.A",
        "bottomB": "C.B.B",
      },
      "middleC": Object {
        "bottomA": "C.C.A",
        "bottomB": "C.C.B",
      },
    },
  },
}
`;

exports[`create function ignores already existed files/directories that is not mismatched 1`] = `
Object {
  "firstBorn": Object {
    "flat": Object {
      "all": Array [
        "tmp/create.1/bar",
        "tmp/create.1/baz",
        "tmp/create.1/emptyBar",
        "tmp/create.1/emptyBaz",
        "tmp/create.1/emptyFoo",
        "tmp/create.1/emptyQux",
        "tmp/create.1/firstBornFiles",
        "tmp/create.1/firstBornFiles/bar",
        "tmp/create.1/firstBornFiles/baz",
        "tmp/create.1/firstBornFiles/foo",
        "tmp/create.1/firstBornFiles/qux",
        "tmp/create.1/foo",
        "tmp/create.1/qux",
      ],
      "directories": Array [
        "tmp/create.1/emptyBar",
        "tmp/create.1/emptyBaz",
        "tmp/create.1/emptyFoo",
        "tmp/create.1/emptyQux",
        "tmp/create.1/firstBornFiles",
      ],
      "fileContents": Object {
        "tmp/create.1/bar": "bar",
        "tmp/create.1/baz": "baz",
        "tmp/create.1/firstBornFiles/bar": "bar",
        "tmp/create.1/firstBornFiles/baz": "baz",
        "tmp/create.1/firstBornFiles/foo": "foo",
        "tmp/create.1/firstBornFiles/qux": "qux",
        "tmp/create.1/foo": "foo",
        "tmp/create.1/qux": "qux",
      },
      "files": Array [
        "tmp/create.1/bar",
        "tmp/create.1/baz",
        "tmp/create.1/firstBornFiles/bar",
        "tmp/create.1/firstBornFiles/baz",
        "tmp/create.1/firstBornFiles/foo",
        "tmp/create.1/firstBornFiles/qux",
        "tmp/create.1/foo",
        "tmp/create.1/qux",
      ],
    },
    "nested": Object {
      "bar": "bar",
      "baz": "baz",
      "emptyBar": Object {},
      "emptyBaz": Object {},
      "emptyFoo": Object {},
      "emptyQux": Object {},
      "firstBornFiles": Object {
        "bar": "bar",
        "baz": "baz",
        "foo": "foo",
        "qux": "qux",
      },
      "foo": "foo",
      "qux": "qux",
    },
  },
  "secondBorn": Object {
    "flat": Object {
      "all": Array [
        "tmp/create.1/bar",
        "tmp/create.1/baz",
        "tmp/create.1/emptyBar",
        "tmp/create.1/emptyBaz",
        "tmp/create.1/emptyFoo",
        "tmp/create.1/emptyQux",
        "tmp/create.1/firstBornFiles",
        "tmp/create.1/firstBornFiles/bar",
        "tmp/create.1/firstBornFiles/baz",
        "tmp/create.1/firstBornFiles/foo",
        "tmp/create.1/firstBornFiles/qux",
        "tmp/create.1/foo",
        "tmp/create.1/newFileA",
        "tmp/create.1/newFileB",
        "tmp/create.1/newFileC",
        "tmp/create.1/newFileD",
        "tmp/create.1/newFileE",
        "tmp/create.1/newTreeA",
        "tmp/create.1/newTreeB",
        "tmp/create.1/newTreeB/empty",
        "tmp/create.1/newTreeC",
        "tmp/create.1/newTreeC/empty",
        "tmp/create.1/newTreeC/emptyFile",
        "tmp/create.1/newTreeC/emptyFile/file",
        "tmp/create.1/newTreeC/nonEmptyFile",
        "tmp/create.1/newTreeC/nonEmptyFile/file",
        "tmp/create.1/newTreeD",
        "tmp/create.1/newTreeD/newTreeA",
        "tmp/create.1/newTreeD/newTreeB",
        "tmp/create.1/newTreeD/newTreeB/empty",
        "tmp/create.1/newTreeD/newTreeC",
        "tmp/create.1/newTreeD/newTreeC/empty",
        "tmp/create.1/newTreeD/newTreeC/emptyFile",
        "tmp/create.1/newTreeD/newTreeC/emptyFile/file",
        "tmp/create.1/newTreeD/newTreeC/nonEmptyFile",
        "tmp/create.1/newTreeD/newTreeC/nonEmptyFile/file",
        "tmp/create.1/qux",
      ],
      "directories": Array [
        "tmp/create.1/emptyBar",
        "tmp/create.1/emptyBaz",
        "tmp/create.1/emptyFoo",
        "tmp/create.1/emptyQux",
        "tmp/create.1/firstBornFiles",
        "tmp/create.1/newTreeA",
        "tmp/create.1/newTreeB",
        "tmp/create.1/newTreeB/empty",
        "tmp/create.1/newTreeC",
        "tmp/create.1/newTreeC/empty",
        "tmp/create.1/newTreeC/emptyFile",
        "tmp/create.1/newTreeC/nonEmptyFile",
        "tmp/create.1/newTreeD",
        "tmp/create.1/newTreeD/newTreeA",
        "tmp/create.1/newTreeD/newTreeB",
        "tmp/create.1/newTreeD/newTreeB/empty",
        "tmp/create.1/newTreeD/newTreeC",
        "tmp/create.1/newTreeD/newTreeC/empty",
        "tmp/create.1/newTreeD/newTreeC/emptyFile",
        "tmp/create.1/newTreeD/newTreeC/nonEmptyFile",
      ],
      "fileContents": Object {
        "tmp/create.1/bar": "bar",
        "tmp/create.1/baz": "baz",
        "tmp/create.1/firstBornFiles/bar": "bar",
        "tmp/create.1/firstBornFiles/baz": "baz",
        "tmp/create.1/firstBornFiles/foo": "foo",
        "tmp/create.1/firstBornFiles/qux": "qux",
        "tmp/create.1/foo": "foo",
        "tmp/create.1/newFileA": "First New File",
        "tmp/create.1/newFileB": "Second New File",
        "tmp/create.1/newFileC": "Third New File",
        "tmp/create.1/newFileD": "Forth New File",
        "tmp/create.1/newFileE": "Fifth New File",
        "tmp/create.1/newTreeC/emptyFile/file": "",
        "tmp/create.1/newTreeC/nonEmptyFile/file": "Not empty",
        "tmp/create.1/newTreeD/newTreeC/emptyFile/file": "",
        "tmp/create.1/newTreeD/newTreeC/nonEmptyFile/file": "Not empty",
        "tmp/create.1/qux": "qux",
      },
      "files": Array [
        "tmp/create.1/bar",
        "tmp/create.1/baz",
        "tmp/create.1/firstBornFiles/bar",
        "tmp/create.1/firstBornFiles/baz",
        "tmp/create.1/firstBornFiles/foo",
        "tmp/create.1/firstBornFiles/qux",
        "tmp/create.1/foo",
        "tmp/create.1/newFileA",
        "tmp/create.1/newFileB",
        "tmp/create.1/newFileC",
        "tmp/create.1/newFileD",
        "tmp/create.1/newFileE",
        "tmp/create.1/newTreeC/emptyFile/file",
        "tmp/create.1/newTreeC/nonEmptyFile/file",
        "tmp/create.1/newTreeD/newTreeC/emptyFile/file",
        "tmp/create.1/newTreeD/newTreeC/nonEmptyFile/file",
        "tmp/create.1/qux",
      ],
    },
    "nested": Object {
      "bar": "bar",
      "baz": "baz",
      "emptyBar": Object {},
      "emptyBaz": Object {},
      "emptyFoo": Object {},
      "emptyQux": Object {},
      "firstBornFiles": Object {
        "bar": "bar",
        "baz": "baz",
        "foo": "foo",
        "qux": "qux",
      },
      "foo": "foo",
      "newFileA": "First New File",
      "newFileB": "Second New File",
      "newFileC": "Third New File",
      "newFileD": "Forth New File",
      "newFileE": "Fifth New File",
      "newTreeA": Object {},
      "newTreeB": Object {
        "empty": Object {},
      },
      "newTreeC": Object {
        "empty": Object {},
        "emptyFile": Object {
          "file": "",
        },
        "nonEmptyFile": Object {
          "file": "Not empty",
        },
      },
      "newTreeD": Object {
        "newTreeA": Object {},
        "newTreeB": Object {
          "empty": Object {},
        },
        "newTreeC": Object {
          "empty": Object {},
          "emptyFile": Object {
            "file": "",
          },
          "nonEmptyFile": Object {
            "file": "Not empty",
          },
        },
      },
      "qux": "qux",
    },
  },
}
`;

exports[`create function throws error when encounter mismatched entities in which a file is requested in place of a folder 1`] = `[Error: EISDIR: illegal operation on a directory, open 'tmp/create.2.0/mismatched']`;

exports[`create function throws error when encounter mismatched entities in which a folder is requested in place of a file 1`] = `[Error: Entity tmp/create.2.1/mismatched exists but is not directory]`;

exports[`traverse function does not dig names end with "B" 1`] = `
Array [
  Object {
    "container": "tmp/traverse.0",
    "item": "topA",
    "level": 0,
    "path": "tmp/traverse.0/topA",
  },
  Object {
    "container": "tmp/traverse.0",
    "item": "topB",
    "level": 0,
    "path": "tmp/traverse.0/topB",
  },
  Object {
    "container": "tmp/traverse.0",
    "item": "topC",
    "level": 0,
    "path": "tmp/traverse.0/topC",
  },
  Object {
    "container": "tmp/traverse.0/topC",
    "item": "middleA",
    "level": 1,
    "path": "tmp/traverse.0/topC/middleA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleA",
    "item": "bottomA",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleA/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleA",
    "item": "bottomB",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleA/bottomB",
  },
  Object {
    "container": "tmp/traverse.0/topC",
    "item": "middleB",
    "level": 1,
    "path": "tmp/traverse.0/topC/middleB",
  },
  Object {
    "container": "tmp/traverse.0/topC",
    "item": "middleC",
    "level": 1,
    "path": "tmp/traverse.0/topC/middleC",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleC",
    "item": "bottomA",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleC/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleC",
    "item": "bottomB",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleC/bottomB",
  },
]
`;

exports[`traverse function only dig names end with "B" 1`] = `
Array [
  Object {
    "container": "tmp/traverse.0",
    "item": "topA",
    "level": 0,
    "path": "tmp/traverse.0/topA",
  },
  Object {
    "container": "tmp/traverse.0",
    "item": "topB",
    "level": 0,
    "path": "tmp/traverse.0/topB",
  },
  Object {
    "container": "tmp/traverse.0/topB",
    "item": "middleA",
    "level": 1,
    "path": "tmp/traverse.0/topB/middleA",
  },
  Object {
    "container": "tmp/traverse.0/topB",
    "item": "middleB",
    "level": 1,
    "path": "tmp/traverse.0/topB/middleB",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleB",
    "item": "bottomA",
    "level": 2,
    "path": "tmp/traverse.0/topB/middleB/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleB",
    "item": "bottomB",
    "level": 2,
    "path": "tmp/traverse.0/topB/middleB/bottomB",
  },
  Object {
    "container": "tmp/traverse.0/topB",
    "item": "middleC",
    "level": 1,
    "path": "tmp/traverse.0/topB/middleC",
  },
  Object {
    "container": "tmp/traverse.0",
    "item": "topC",
    "level": 0,
    "path": "tmp/traverse.0/topC",
  },
]
`;

exports[`traverse function works with default \`deep\` and \`level\` 1`] = `
Array [
  Object {
    "container": "tmp/traverse.0",
    "item": "topA",
    "level": 0,
    "path": "tmp/traverse.0/topA",
  },
  Object {
    "container": "tmp/traverse.0",
    "item": "topB",
    "level": 0,
    "path": "tmp/traverse.0/topB",
  },
  Object {
    "container": "tmp/traverse.0/topB",
    "item": "middleA",
    "level": 1,
    "path": "tmp/traverse.0/topB/middleA",
  },
  Object {
    "container": "tmp/traverse.0/topB",
    "item": "middleB",
    "level": 1,
    "path": "tmp/traverse.0/topB/middleB",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleB",
    "item": "bottomA",
    "level": 2,
    "path": "tmp/traverse.0/topB/middleB/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleB",
    "item": "bottomB",
    "level": 2,
    "path": "tmp/traverse.0/topB/middleB/bottomB",
  },
  Object {
    "container": "tmp/traverse.0/topB",
    "item": "middleC",
    "level": 1,
    "path": "tmp/traverse.0/topB/middleC",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleC",
    "item": "bottomA",
    "level": 2,
    "path": "tmp/traverse.0/topB/middleC/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleC",
    "item": "bottomB",
    "level": 2,
    "path": "tmp/traverse.0/topB/middleC/bottomB",
  },
  Object {
    "container": "tmp/traverse.0",
    "item": "topC",
    "level": 0,
    "path": "tmp/traverse.0/topC",
  },
  Object {
    "container": "tmp/traverse.0/topC",
    "item": "middleA",
    "level": 1,
    "path": "tmp/traverse.0/topC/middleA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleA",
    "item": "bottomA",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleA/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleA",
    "item": "bottomB",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleA/bottomB",
  },
  Object {
    "container": "tmp/traverse.0/topC",
    "item": "middleB",
    "level": 1,
    "path": "tmp/traverse.0/topC/middleB",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleB",
    "item": "bottomA",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleB/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleB",
    "item": "bottomB",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleB/bottomB",
  },
  Object {
    "container": "tmp/traverse.0/topC",
    "item": "middleC",
    "level": 1,
    "path": "tmp/traverse.0/topC/middleC",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleC",
    "item": "bottomA",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleC/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleC",
    "item": "bottomB",
    "level": 2,
    "path": "tmp/traverse.0/topC/middleC/bottomB",
  },
]
`;

exports[`traverse function works with provided \`deep\` and \`level\` 1`] = `
Array [
  Object {
    "container": "tmp/traverse.0",
    "item": "topA",
    "level": 3,
    "path": "tmp/traverse.0/topA",
  },
  Object {
    "container": "tmp/traverse.0",
    "item": "topB",
    "level": 3,
    "path": "tmp/traverse.0/topB",
  },
  Object {
    "container": "tmp/traverse.0/topB",
    "item": "middleA",
    "level": 4,
    "path": "tmp/traverse.0/topB/middleA",
  },
  Object {
    "container": "tmp/traverse.0/topB",
    "item": "middleB",
    "level": 4,
    "path": "tmp/traverse.0/topB/middleB",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleB",
    "item": "bottomA",
    "level": 5,
    "path": "tmp/traverse.0/topB/middleB/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleB",
    "item": "bottomB",
    "level": 5,
    "path": "tmp/traverse.0/topB/middleB/bottomB",
  },
  Object {
    "container": "tmp/traverse.0/topB",
    "item": "middleC",
    "level": 4,
    "path": "tmp/traverse.0/topB/middleC",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleC",
    "item": "bottomA",
    "level": 5,
    "path": "tmp/traverse.0/topB/middleC/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topB/middleC",
    "item": "bottomB",
    "level": 5,
    "path": "tmp/traverse.0/topB/middleC/bottomB",
  },
  Object {
    "container": "tmp/traverse.0",
    "item": "topC",
    "level": 3,
    "path": "tmp/traverse.0/topC",
  },
  Object {
    "container": "tmp/traverse.0/topC",
    "item": "middleA",
    "level": 4,
    "path": "tmp/traverse.0/topC/middleA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleA",
    "item": "bottomA",
    "level": 5,
    "path": "tmp/traverse.0/topC/middleA/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleA",
    "item": "bottomB",
    "level": 5,
    "path": "tmp/traverse.0/topC/middleA/bottomB",
  },
  Object {
    "container": "tmp/traverse.0/topC",
    "item": "middleB",
    "level": 4,
    "path": "tmp/traverse.0/topC/middleB",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleB",
    "item": "bottomA",
    "level": 5,
    "path": "tmp/traverse.0/topC/middleB/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleB",
    "item": "bottomB",
    "level": 5,
    "path": "tmp/traverse.0/topC/middleB/bottomB",
  },
  Object {
    "container": "tmp/traverse.0/topC",
    "item": "middleC",
    "level": 4,
    "path": "tmp/traverse.0/topC/middleC",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleC",
    "item": "bottomA",
    "level": 5,
    "path": "tmp/traverse.0/topC/middleC/bottomA",
  },
  Object {
    "container": "tmp/traverse.0/topC/middleC",
    "item": "bottomB",
    "level": 5,
    "path": "tmp/traverse.0/topC/middleC/bottomB",
  },
]
`;
